tests/project_helpers.py

```diff
@@ -0,0 +1,245 @@
+
+        The installation script will change if the code to setup the venv on
+        `venv_script` changes or if the requirements field of the experiment
+        changes. It will also be different for different Python versions.
+        The Python version currently used to run the tests is used to compute
+        the hash and create the venv.
+        """
+        dummy_script = self.venv_script(Path("/dev/null"))
+        py_version = f"{sys.version_info.major},{sys.version_info.major}"
+        script_and_version_bytes = ("".join(dummy_script) + py_version).encode()
+        return hashlib.sha256(script_and_version_bytes).hexdigest()
+
+    def __call__(self, cache: pytest.Cache) -> Path:
+        """Get this venv's dir and create it if necessary.
+
+        The venv_dir is where we install the dependencies of the current
+        experiment. It is keyed by the sha256 checksum of the requirements
+        file and the script we use for setting up the venv. This way, we
+        don't risk using a previously cached venv for a different if the
+        script or the requirements to create that venv change.
+        """
+        # We cache venv dirs using the hash from create_venv_hash
+        cached_str = cache.get(f"fawltydeps/{self.venv_hash()}", None)
+        if cached_str is not None and Path(cached_str, ".installed").is_file():
+            return Path(cached_str)  # already cached
+
+        # Must run the script to set up the venv
+        venv_dir = Path(cache.mkdir(f"fawltydeps_venv_{self.venv_hash()}"))
+        logger.info(f"Creating venv at {venv_dir}...")
+        venv_script = self.venv_script(venv_dir)
+        subprocess.run(
+            " && ".join(venv_script),
+            check=True,  # fail if any of the commands fail
+            shell=True,  # pass multiple shell commands to the subprocess
+        )
+        # Make sure the venv has been installed
+        assert Path(venv_dir, ".installed").is_file()
+        cache.set(f"fawltydeps/{self.venv_hash()}", str(venv_dir))
+        return venv_dir
+
+
+@dataclass
+class AnalysisExpectations:
+    """Verify our expectations on the analysis resulting from an experiment.
+
+    Encapsulate the expected values of an Analysis object (or its JSON
+    counterpart) after a real/sample project experiment has been run.
+
+    The members of this expectation object are all optional; if omitted, the
+    corresponding Analysis member will _not_ be checked. Furthermore, no "deep"
+    verification of each member is performed here; rather, we only check that
+    the set of (import or dependency) names that can found in each Analysis
+    member match the expected set stored here.
+    """
+
+    imports: Optional[Set[str]] = None
+    declared_deps: Optional[Set[str]] = None
+    undeclared_deps: Optional[Set[str]] = None
+    unused_deps: Optional[Set[str]] = None
+
+    @classmethod
+    def parse_from_toml(cls, data: TomlData) -> "AnalysisExpectations":
```

Comment:
What's the reason to name and type this as TOML-specific rather than mapping-like? Intuitively, I'd name this `parse_from_mapping` and have `data: Mapping[str, Any]`, but am I missing something?

Reply:
I mean, `TomlData = Dict[str, Any]`, right?

Reply:
What @mknorps said. I want to communicate that this is data from a TOML file (which tells the reader more than a simple `Dict[str, Any]` or equivalent), and I _also_ want to make it easier (in a possible future where the TOML parser itself returns something more strictly typed than `Any`) to narrow the definition of `TomlData` to something stricter.

Reply:
ahhh OK, makes sense. i was trying to generalize / broaden but makes sense the potential desire to instead prepare for the opposite, potential to narrow.